### ISSUE TITLE
MultipartReader$Companion.readHeaders throws Unexpected header

### ISSUE THREAD

**Issue by ramanandsingh85** [apollographql/apollo-kotlin#4596]
### Version

com.apollographql.apollo3: 3.7.3,  graphql-yoga: 3.1.1

### Summary

Graphql Server 3.x is returning invalid multipart response. Following header is added by the android apollo kotlin sdk which seems the sole cause:

-H "Accept:multipart/mixed; deferSpec=20220824, application/json"

It is clear from logs that five hyphens are appended at the end of response which making multipart rader throw exception:

Unexpected header: -----

### Steps to reproduce the behavior

Use any graphql query with the above version set and it will throw an error.

It was working in android kotlin sdk 2.x

### Logs

```
-29946 okhttp.OkHttpClient     staging.xxxx.app                 I  --> POST https://api.xxx-staging.live/graphql
2022-12-22 17:13:12.382 29892-29946 okhttp.OkHttpClient     staging.xxxx.app                 I  Content-Type: application/json
2022-12-22 17:13:12.382 29892-29946 okhttp.OkHttpClient     staging.xxxx.app                 I  Content-Length: 147
2022-12-22 17:13:12.382 29892-29946 okhttp.OkHttpClient     staging.xxxx.app                 I  X-APOLLO-OPERATION-ID: 530c832b59f59d578fbc1e895e4ff9777ccad79dcb221af71bc961da07f3a281
2022-12-22 17:13:12.382 29892-29946 okhttp.OkHttpClient     staging.xxxx.app                 I  X-APOLLO-OPERATION-NAME: tipItemList
2022-12-22 17:13:12.382 29892-29946 okhttp.OkHttpClient     staging.xxxx.app                 I  Accept: multipart/mixed; deferSpec=20220824, application/json
2022-12-22 17:13:12.383 29892-29946 okhttp.OkHttpClient     staging.xxxx.app                 I  {"operationName":"tipItemList","variables":{},"query":"query tipItemList { tipItemList { amount animation_asset_link created_at id image name } }"}
2022-12-22 17:13:12.383 29892-29946 okhttp.OkHttpClient     staging.xxxx.app                 I  --> END POST (147-byte body)
2022-12-22 17:13:12.816 29892-30129 okhttp.OkHttpClient     staging.xxxx.app                 I  <-- 101 Switching Protocols https://api.xxxshop-staging.live/graphql (1008ms)
2022-12-22 17:13:12.817 29892-30129 okhttp.OkHttpClient     staging.xxxx.app                 I  Date: Thu, 22 Dec 2022 11:43:25 GMT
2022-12-22 17:13:12.817 29892-30129 okhttp.OkHttpClient     staging.xxxx.app                 I  Connection: upgrade
2022-12-22 17:13:12.817 29892-30129 okhttp.OkHttpClient     staging.xxxx.app                 I  Upgrade: websocket
2022-12-22 17:13:12.817 29892-30129 okhttp.OkHttpClient     staging.xxxx.app                 I  Sec-WebSocket-Accept: Oi0sweoMmN47sd1ECsMCX4d4UE8=
2022-12-22 17:13:12.817 29892-30129 okhttp.OkHttpClient     staging.xxxx.app                 I  Sec-WebSocket-Protocol: graphql-transport-ws
2022-12-22 17:13:12.818 29892-30129 okhttp.OkHttpClient     staging.xxxx.app                 I  <-- END HTTP
2022-12-22 17:13:12.860 29892-29991 EGL_emulation           staging.xxxx.app                 D  app_time_stats: avg=9.13ms min=1.55ms max=28.66ms count=62
2022-12-22 17:13:12.891 29892-29944 okhttp.OkHttpClient     staging.xxxx.app                 I  <-- 200 https://api.xxxshop-staging.live/graphql (1194ms)
2022-12-22 17:13:12.893 29892-29944 okhttp.OkHttpClient     staging.xxxx.app                 I  date: Thu, 22 Dec 2022 11:43:25 GMT
2022-12-22 17:13:12.893 29892-29944 okhttp.OkHttpClient     staging.xxxx.app                 I  content-type: multipart/mixed; boundary="-"
2022-12-22 17:13:12.893 29892-29944 okhttp.OkHttpClient     staging.xxxx.app                 I  xxx-trace-id: ead93dc49d6845bd196680ddf88506f9
2022-12-22 17:13:12.893 29892-29944 okhttp.OkHttpClient     staging.xxxx.app                 I  access-control-allow-origin: *
2022-12-22 17:13:12.894 29892-29944 okhttp.OkHttpClient     staging.xxxx.app                 I  ---
2022-12-22 17:13:12.894 29892-29944 okhttp.OkHttpClient     staging.xxxx.app                 I  Content-Type: application/json; charset=utf-8
2022-12-22 17:13:12.894 29892-29944 okhttp.OkHttpClient     staging.xxxx.app                 I  Content-Length: 776
2022-12-22 17:13:12.894 29892-29944 okhttp.OkHttpClient     staging.xxxx.app                 I  
2022-12-22 17:13:12.894 29892-29944 okhttp.OkHttpClient     staging.xxxx.app                 I  {"data":{"stream":{"id":1080,"title":"first testing stream","is_live":true,"likes":226,"isInWatchlist":false,"stream_recordings":null,"preview_image":"https://cdn.xxxshop.live/QPd0STKNf26mDKW4sCbO4.jpg","preview_video":null,"rtc_token":"006afd07411e1914563bbe3094f7b818518IADazGcAnvV3UhiNgi6llpXK54yopNXwAAC3wj5SaKunlkZaLZcUergdEADRzuCOXZSlYwEAAQBdlKVj","start_datetime":"2022-06-21T06:03:00.000Z","stream_slug":"stream1","watchlist_count":1,"went_live_at":"2022-09-06T04:58:12.515Z","went_offline_at":"2022-12-21T20:46:31.601Z","end_datetime":"2022-12-22T11:46:46.780Z","host":{"id":10633,"name":"Bhavin","slug":"bhavinSavsani","picture":"https://cdn.xxxshop.live/user-profile/7pjm83Bu3k_wt2FEftdCP.jpg","is_self_following":true,"nickname":"bhavinSavsani","ratings":4.3}}}}
2022-12-22 17:13:12.894 29892-29944 okhttp.OkHttpClient     staging.xxxx.app                 I  ---
2022-12-22 17:13:12.894 29892-29944 okhttp.OkHttpClient     staging.xxxx.app                 I  -----
2022-12-22 17:13:12.895 29892-29944 okhttp.OkHttpClient     staging.xxxx.app                 I  <-- END HTTP (865-byte body)
2022-12-22 17:13:12.904 29892-29892 System.err              staging.xxxx.app                 W  java.lang.IllegalStateException: Unexpected header: -----
2022-12-22 17:13:12.905 29892-29892 System.err              staging.xxxx.app                 W  	at com.apollographql.apollo3.internal.MultipartReader$Companion.readHeaders(MultipartReader.kt:206)
2022-12-22 17:13:12.905 29892-29892 System.err              staging.xxxx.app                 W  	at com.apollographql.apollo3.internal.MultipartReader$Companion.access$readHeaders(MultipartReader.kt:190)
2022-12-22 17:13:12.905 29892-29892 System.err              staging.xxxx.app                 W  	at com.apollographql.apollo3.internal.MultipartReader.nextPart(MultipartReader.kt:136)
2022-12-22 17:13:12.906 29892-29892 System.err              staging.xxxx.app                 W  	at com.apollographql.apollo3.internal.MultipartKt$multipartBodyFlow$1.invokeSuspend(multipart.kt:27)
2022-12-22 17:13:12.906 29892-29892 System.err              staging.xxxx.app                 W  	at com.apollographql.apollo3.internal.MultipartKt$multipartBodyFlow$1.invoke(Unknown Source:8)
2022-12-22 17:13:12.906 29892-29892 System.err              staging.xxxx.app                 W  	at com.apollographql.apollo3.internal.MultipartKt$multipartBodyFlow$1.invoke(Unknown Source:4)
2022-12-22 17:13:12.906 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.flow.SafeFlow.collectSafely(Builders.kt:61)
2022-12-22 17:13:12.907 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.flow.AbstractFlow.collect(Flow.kt:230)
2022-12-22 17:13:12.907 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.flow.FlowKt__EmittersKt$onCompletion$$inlined$unsafeFlow$1.collect(SafeCollector.common.kt:114)
2022-12-22 17:13:12.907 29892-29892 System.err              staging.xxxx.app                 W  	at com.apollographql.apollo3.network.http.HttpNetworkTransport$multipleResponses$$inlined$map$1.collect(SafeCollector.common.kt:113)
2022-12-22 17:13:12.907 29892-29892 System.err              staging.xxxx.app                 W  	at com.apollographql.apollo3.network.http.HttpNetworkTransport$multipleResponses$$inlined$filterNot$1.collect(SafeCollector.common.kt:113)
2022-12-22 17:13:12.907 29892-29892 System.err              staging.xxxx.app                 W  	at com.apollographql.apollo3.network.http.HttpNetworkTransport$execute$1$invokeSuspend$$inlined$map$1.collect(SafeCollector.common.kt:113)
2022-12-22 17:13:12.907 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.flow.FlowKt__CollectKt.emitAll(Collect.kt:109)
2022-12-22 17:13:12.907 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.flow.FlowKt.emitAll(Unknown Source:1)
2022-12-22 17:13:12.907 29892-29892 System.err              staging.xxxx.app                 W  	at com.apollographql.apollo3.network.http.HttpNetworkTransport$execute$1.invokeSuspend(HttpNetworkTransport.kt:78)
2022-12-22 17:13:12.907 29892-29892 System.err              staging.xxxx.app                 W  	at com.apollographql.apollo3.network.http.HttpNetworkTransport$execute$1.invoke(Unknown Source:8)
2022-12-22 17:13:12.907 29892-29892 System.err              staging.xxxx.app                 W  	at com.apollographql.apollo3.network.http.HttpNetworkTransport$execute$1.invoke(Unknown Source:4)
2022-12-22 17:13:12.907 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.flow.SafeFlow.collectSafely(Builders.kt:61)
2022-12-22 17:13:12.907 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.flow.AbstractFlow.collect(Flow.kt:230)
2022-12-22 17:13:12.907 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.flow.internal.ChannelFlowOperatorImpl.flowCollect(ChannelFlow.kt:195)
2022-12-22 17:13:12.907 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.flow.internal.ChannelFlowOperator.collectTo$suspendImpl(ChannelFlow.kt:157)
2022-12-22 17:13:12.908 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.flow.internal.ChannelFlowOperator.collectTo(Unknown Source:0)
2022-12-22 17:13:12.908 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.flow.internal.ChannelFlow$collectToFun$1.invokeSuspend(ChannelFlow.kt:60)
2022-12-22 17:13:12.908 29892-29892 System.err              staging.xxxx.app                 W  	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
2022-12-22 17:13:12.908 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
2022-12-22 17:13:12.908 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
2022-12-22 17:13:12.908 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
2022-12-22 17:13:12.908 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
2022-12-22 17:13:12.908 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
2022-12-22 17:13:12.908 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
2022-12-22 17:13:12.908 29892-29892 System.err              staging.xxxx.app                 W  	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
2022-12-22 17:13:12.916 29892
```

**Comment by BoD** [comment]
Thank you for reporting this!

This seems to stem from an incompatibility with the last empty part of the multipart response sent by the server. We'll look into make the code more lenient for this case, but in the meantime, you can make Apollo not request multipart at all, like so:

```kotlin
    val client = ApolloClient.Builder()
        // ...
        .addHttpInterceptor(object : HttpInterceptor {
          override suspend fun intercept(request: HttpRequest, chain: HttpInterceptorChain): HttpResponse {
            return chain.proceed(request.newBuilder().headers(request.headers.filterNot { it.name.equals("accept", ignoreCase = true) } + HttpHeader("Accept", "application/json")).build())
          }
        })
        .build()
```

**Comment by ramanandsingh85** [comment]
Thank you for your quick response,

Definitely answer is helpful. Good to go for now.

**Comment by martinbonnin** [comment]
This is now available in 3.7.4
